### PR TITLE
Legacy generic build: Test BF_X2 = 3 if BF_X2 = 1 was faster than BF_X2 = 0

### PR DESCRIPTION
**Pull request by frank-dittrich**
I think a similar change should be in core. As long as this isn't the case, I'd like to have it in bleeding-jumbo.

### CODE REVIEW

File: src/Makefile.legacy [hunk 1]
```diff
@@ -1904,10 +1904,10 @@ bench-t.o: bench.c
 	@
 
 fmt_externs.h: $(PLUGFORMATS_SRCS) Makefile.legacy *_fmt_plug.c
-	$(CC) -E -P $(CFLAGS) -DFMT_EXTERNS_H *_fmt_plug.c | LC_ALL=C $(GREP) "extern struct fmt_main" | LC_ALL=C $(SORT) > fmt_externs.h
+	$(CC) -E -P $(CFLAGS) -DFMT_EXTERNS_H *_fmt_plug.c | LC_ALL=C $(GREP) "extern struct fmt_main" | LC_ALL=C $(SORT)-f > fmt_externs.h
```
Comment (magnumripper) [inline src/Makefile.legacy]: `$(SORT)-f` I think you are lacking a space there?

Comment (frank-dittrich) [inline src/Makefile.legacy]: will fix, sorry for not testing this change

File: src/Makefile.legacy [hunk 2]
```diff
@@ -1759,10 +1757,6 @@ $(SUBDIRS):
 ../run/unique.com: john.com
 	copy john.com ..\run\unique.com
 
-# This must not be specified with path because "make clean" would delete it.
-john.local.conf:
-	[ -f ../run/john.local.conf ] || touch ../run/john.local.conf
-
```
Comment (magnumripper) [inline src/Makefile.legacy]: Wow, how could we miss this?! That was long ago!

Comment (frank-dittrich) [inline src/Makefile.legacy]: I just compared 2 different clones. after I used the first one for an autoconf build and `make -s distclean` and the second one for a legacy build and `make -s -f Makefile.legacy clean`.

The other difference between those 2 clones is an empty `run/kernels` directory in the autoconf clone. 
Apparently, this directory is created even for CPU-only builds.
(May be we should always remove run/kernels on `make distclean`, but don't create it for CPU builds.)

Comment (frank-dittrich) [inline src/Makefile.legacy]: BTW: Autoconf isn't just creating an *empty* run/kernels directory for non-OpenCL-builds!
Instead, it contains 96 files (*.h and *.cl).
This is somewhat unexpected.

**Comment by frank-dittrich** [comment]
These are the known issued with legacy builds which still remain:

The following binaries (or symlinks) are not created in a legecy build:
```
run/bitlocker2john
run/dmg2john
run/eapmd5tojohn
run/keepass2john
run/putty2john
```

These warnings occur in a linux-x86-64 legacy build, but not in a linux-x86-64-native or generic legacy build (AVX cpu):
```
$ make -s -f Makefile.legacy linux-x86-64  
stribog_fmt_plug.c:501:2: warning: #warning Stribog-256 and Stribog-512 formats require SSE 4.1, formats disabled [-Wcpp]
 #warning Stribog-256 and Stribog-512 formats require SSE 4.1, formats disabled
  ^
ar: creating aes.a
In file included from secp256k1.c:13:0:
group_impl.h:129:13: warning: 'secp256k1_ge_set_all_gej_var' defined but not used [-Wunused-function]
 static void secp256k1_ge_set_all_gej_var(secp256k1_ge *r, const secp256k1_gej *a, size_t len, const secp256k1_callback *cb) {
             ^
In file included from secp256k1.c:13:0:
group_impl.h:686:12: warning: 'secp256k1_gej_has_quad_y_var' defined but not used [-Wunused-function]
 static int secp256k1_gej_has_quad_y_var(const secp256k1_gej *a) {
            ^
group_impl.h:361:13: warning: 'secp256k1_gej_add_var' defined but not used [-Wunused-function]
 static void secp256k1_gej_add_var(secp256k1_gej *r, const secp256k1_gej *a, const secp256k1_gej *b, secp256k1_fe *rzr) {
             ^
In file included from secp256k1.c:13:0:
group_impl.h:113:13: warning: 'secp256k1_ge_set_gej_var' defined but not used [-Wunused-function]
 static void secp256k1_ge_set_gej_var(secp256k1_ge *r, secp256k1_gej *a) {
             ^
In file included from secp256k1.c:13:0:
group_impl.h:267:12: warning: 'secp256k1_gej_is_valid_var' defined but not used [-Wunused-function]
 static int secp256k1_gej_is_valid_var(const secp256k1_gej *a) {
            ^
In file included from secp256k1.c:14:0:
ecmult_impl.h:216:12: warning: 'secp256k1_ecmult_context_is_built' defined but not used [-Wunused-function]
 static int secp256k1_ecmult_context_is_built(const secp256k1_ecmult_context *ctx) {
            ^
In file included from secp256k1.c:16:0:
ecmult_gen_impl.h:94:12: warning: 'secp256k1_ecmult_gen_context_is_built' defined but not used [-Wunused-function]
 static int secp256k1_ecmult_gen_context_is_built(const secp256k1_ecmult_gen_context* ctx) {
            ^
ar: creating secp256k1.a
Failing after this point just means some helper tools did not build:
All done
```

**Comment by frank-dittrich** [comment]
Another problem I noticed when comparing legacy generic and legacy linux-x86-64-native, but this may be unrelated to legacy:

For these formats, there's  no warning when they are disabled due to preprocessor commands:
```
Raw-SHA1-ng
Raw-SHA256-ng
Raw-SHA512-ng
```

There should be a warning similar to the one for these missing formats in a legacy generic build:
```
Stribog-256
Stribog-512
```

(There's no warning to the missing crypt format in legacy generic, but a core generic build also has no crypt format support and doesn't warn about it.)

**Comment by frank-dittrich** [comment]
BTW: Out of 374 formats in a legacy generic build, these are the 70 formats where the password length differs, compared to a legacy linux-x86-64-avx build:
```
7z	41 [worst case UTF-8] to 125 [ASCII]	28
dynamic=md5($p)	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_0	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_1	26 [worst case UTF-8] to 78 [ASCII]	7 [worst case UTF-8] to 23 [ASCII]
dynamic_10	29 [worst case UTF-8] to 87 [ASCII]	10 [worst case UTF-8] to 32 [ASCII]
dynamic_1001	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_1002	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_1004	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_1005	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_1006	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_1007	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_1008	31 [worst case UTF-8] to 94 [ASCII]	13 [worst case UTF-8] to 39 [ASCII]
dynamic_1009	31 [worst case UTF-8] to 94 [ASCII]	13 [worst case UTF-8] to 39 [ASCII]
dynamic_1011	16 [worst case UTF-8] to 48 [ASCII]	7 [worst case UTF-8] to 23 [ASCII]
dynamic_1012	16 [worst case UTF-8] to 48 [ASCII]	7 [worst case UTF-8] to 23 [ASCII]
dynamic_1013	16 [worst case UTF-8] to 48 [ASCII]	7 [worst case UTF-8] to 23 [ASCII]
dynamic_1015	18 [worst case UTF-8] to 56 [ASCII]	10 [worst case UTF-8] to 31 [ASCII]
dynamic_1018	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_1019	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_1020	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_1021	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_1022	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_1024	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_1025	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_1034	26 [worst case UTF-8] to 78 [ASCII]	7 [worst case UTF-8] to 23 [ASCII]
dynamic_11	29 [worst case UTF-8] to 87 [ASCII]	10 [worst case UTF-8] to 32 [ASCII]
dynamic_1300	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_1350	36 [worst case UTF-8] to 108 [ASCII]	17 [worst case UTF-8] to 53 [ASCII]
dynamic_14	26 [worst case UTF-8] to 80 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_1400	41 [worst case UTF-8] to 110 [ASCII]	41 [worst case UTF-8] to 55 [ASCII]
dynamic_1401	36 [worst case UTF-8] to 110 [ASCII]	7 [worst case UTF-8] to 23 [ASCII]
dynamic_1506	36 [worst case UTF-8] to 110 [ASCII]	7 [worst case UTF-8] to 23 [ASCII]
dynamic_1550	36 [worst case UTF-8] to 110 [ASCII]	7 [worst case UTF-8] to 23 [ASCII]
dynamic_1551	36 [worst case UTF-8] to 110 [ASCII]	7 [worst case UTF-8] to 23 [ASCII]
dynamic_2	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_22	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_23	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_29	40	27
dynamic_3	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_30	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_31	28 [worst case UTF-8] to 86 [ASCII]	10 [worst case UTF-8] to 31 [ASCII]
dynamic_32	28 [worst case UTF-8] to 86 [ASCII]	10 [worst case UTF-8] to 31 [ASCII]
dynamic_33	40	27
dynamic_34	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_4	28 [worst case UTF-8] to 86 [ASCII]	10 [worst case UTF-8] to 31 [ASCII]
dynamic_5	18 [worst case UTF-8] to 56 [ASCII]	10 [worst case UTF-8] to 31 [ASCII]
dynamic_6	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
dynamic_8	26 [worst case UTF-8] to 78 [ASCII]	7 [worst case UTF-8] to 23 [ASCII]
dynamic_9	26 [worst case UTF-8] to 80 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
EPiServer	32	19
leet	41 [worst case UTF-8] to 125 [ASCII]	26 [worst case UTF-8] to 79 [ASCII]
MediaWiki	26 [worst case UTF-8] to 80 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
MSCHAPv2	41 [worst case UTF-8] to 64 [ASCII]	27
netntlm	41 [worst case UTF-8] to 64 [ASCII]	27
NT	41 [worst case UTF-8] to 125 [ASCII]	27
osc	28 [worst case UTF-8] to 86 [ASCII]	10 [worst case UTF-8] to 31 [ASCII]
PHPS	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
PHPS2	36 [worst case UTF-8] to 110 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
rar	41 [worst case UTF-8] to 125 [ASCII]	26
Raw-MD4	41 [worst case UTF-8] to 125 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
Raw-MD5	41 [worst case UTF-8] to 125 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
Raw-MD5u	41 [worst case UTF-8] to 125 [ASCII]	27
Raw-SHA1	41 [worst case UTF-8] to 125 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
Raw-SHA1-AxCrypt	41 [worst case UTF-8] to 125 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
Raw-SHA1-Linkedin	41 [worst case UTF-8] to 125 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
Raw-SHA224	41 [worst case UTF-8] to 125 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
Raw-SHA256	41 [worst case UTF-8] to 125 [ASCII]	18 [worst case UTF-8] to 55 [ASCII]
Raw-SHA384	41 [worst case UTF-8] to 125 [ASCII]	37 [worst case UTF-8] to 111 [ASCII]
Raw-SHA512	41 [worst case UTF-8] to 125 [ASCII]	37 [worst case UTF-8] to 111 [ASCII]
saph	13 [worst case UTF-8] to 40 [ASCII]	7 [worst case UTF-8] to 23 [ASCII]
```

**Comment by magnumripper** [comment]
>For these formats, there's  no warning when they are disabled due to preprocessor commands:
```
Raw-SHA1-ng
Raw-SHA256-ng
Raw-SHA512-ng
```
>
>There should be a warning similar to the one for these missing formats in a legacy generic build:
```
Stribog-256
Stribog-512
```

I disagree (if I get you right). The three first formats are merely variants of existing formats in a generic build. The Stribog ones are unique, no?

**Comment by frank-dittrich** [comment]
@magnumripper You are right regarding the RAW-SHA*-ng formats.

**Comment by frank-dittrich** [comment]
SInce @kholia is clearly opposed to spend effort on legacy builds, see
https://github.com/magnumripper/JohnTheRipper/pull/2687#issuecomment-323561349, I'd say I am OK with dropping legacy build support if we can somehow integrate the few advantages a legacy generic build has into the autoconf build.
The resulting build could even be better than a legacy generic build.

Advantages:
-improved max. password length for some builds, but with a performance penalty
-much better bcrypt performance on some systems (with my commit https://github.com/magnumripper/JohnTheRipper/pull/2689/commits/387fc7e91c97d77a60190f924c3695218f7896e8)

Would a special case handling for `./configure --build=generic` or  `./configure --host=generic` be possible? Can we integrate the legacy best.sh performance test into it?

Alternatively, can we allow an option to compare bcrypt performance for BF_X2 = 1 vs. BF_X2 = 3 and pick the better setting?
And implement a setting which allows longer passwords (up to each format's real life max. length), but with a performance penalty?
(Supporting this has probably to be done in each format individually, by allowing 2 iterations instead of 1 depending on a global setting, etc. And it may require a format version change, including new fields to distinguish max. password length in bytes and max. password length in characters.)

**Comment by magnumripper** [comment]
@frank-dittrich I'm OK with all three commits as long as Solar doesn't see some problem with the BF stuff. We now have a "This branch has conflicts that must be resolved" though.

**Comment by kholia** [comment]
My two recent changes to `Makefile.legacy` must have caused these conflicts. They should be easy to resolve though.

**Comment by frank-dittrich** [comment]
No problem, I can rebase and push --force to my repo.
But not today (too busy).

**Comment by kholia** [comment]
The unused function warnings should be now gone from the `secp256k1` code.

**Pull request merged**